Выполнила ДЗ №4

alisaaniskina opened the pull request.

Commit:
Добавила основную программу и unit-test`ы

Commit:
add .travis.yml

Commit:
add status image

Commit:
Намеренно сделала ошибку в 9 строке

в 9 строке сделала ошибку
в 8 строке закомментирован верный вариант

Commit:
Исправила отображение бейджа

Commit:
Исправила сделанную ошибку на верный вариант

Раскомментировала верную 8 строку
Закомментировала 9 строку с ошибкой

skyline-gleb commented on README.md:
```diff
@@ -1,4 +1,7 @@
 # ProjectT spring-2016 - Домашнее задание #4
+
+[![Build Status](https://travis-ci.org/alisaaniskina/lesson_4_homework.svg?branch=lion_final_state_machine)](https://travis-ci.org/alisaaniskina/lesson_4_homework)
```
изменения хотят попасть в мастер, поэтому тут есть ошибка

skyline-gleb commented on mainLion.py:
```diff
@@ -0,0 +1,41 @@
+class Lion :
+    def __init__(self, x, y) :
```
при создании класса `Lion` как понять что я должен передать в констурктор? Плохой нейминг переменных.

skyline-gleb commented on mainLion.py:
```diff
@@ -0,0 +1,41 @@
+    def __init__(self, x, y) :
+        self.state = x
+        self.action = ""
+        self.read_and_go(y)
```
вообще инициализация врятли подразумевает и действие, сразу. Точно так стоит делать?

skyline-gleb commented on test.py:
```diff
@@ -0,0 +1,41 @@
+
+class LionTest (TestCase) :
+    def test_init (self) :
+        res = Lion(1, "hunter")
```
`res` ? результат? а кажется это конечный автомат Лев

skyline-gleb commented:
Протестирован констурктор, а у льва есть еще куча функций, которые как бы публичные, и доступны пользователю. Их тоже надо проверить. А вот тогда и надо бы задуматься, а правильно ли спроектирован конечный автомат, все ли эти функции должны быть публичные?

skyline-gleb commented:
Давай сначала подумаем над самим Львом, а потом уже займешся тестами.

Commit:
Исправила отображение бейджа

Commit:
Исправила конечный автомат

Исправила нейминг переменных
Исправила инициализацию
Сделала функции приватными

Commit:
Исправила тесты

Разделила по категориям
Добавила проверку инициализации
Добавила проверку на некорректный объект

skyline-gleb commented on mainLion.py:
```diff
@@ -0,0 +1,39 @@
+            self.state = state
+        else:
+            raise ValueError("Incorrect state")
+        self.action = ""
```
тут хорошей практикой является заинвертить условие, что в самом начале и сразу кинуть ошибку. тем самым остальной код будет написан дальше, уже без лишнего отступа, легче читаем и все такое

skyline-gleb commented on test.py:
```diff
@@ -0,0 +1,63 @@
+
+    def test_incorrect_arguments (self) :
+        self.assertRaises(ValueError, Lion, "happy")
+        self.assertRaises(ValueError, Lion, "melancholy")
```
вторая проверка она излишняя, так как эквивалентная первой, а лишнюю работу делать плохо

skyline-gleb commented:
И так как надо пару исправлений сделать, еще вдобавок улучши имена тестов на конструктор, чтоб было понятно что они его проверяют, а то, например, `test_full` ну не говорит об этом

Commit:
перенесла обработку ошибки наверх

Commit:
убрала лишнюю проверку, поменяла имена тестов

skyline-gleb commented:
:+1:
